RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorKernels.cu

```diff
@@ -1,416 +1,285 @@
-    unsigned int minHitsPerNtuplet,
-    unsigned int maxNumberOfDoublets)
-{
-  kernelFindNtuplets<<<gridSize, blockSize, 0, cudaStream>>>(
-    cells,
-    nCells,
-    foundNtuplets,
-    minHitsPerNtuplet,
-    maxNumberOfDoublets);
-  cudaCheck(cudaGetLastError());
+
+__global__
+void kernel_VerifyFit(TuplesOnGPU::Container const * __restrict__ tuples,
+                 Rfit::helix_fit const *  __restrict__ fit_results,
+                 Quality *  __restrict__ quality) {
+
+  auto idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx>= tuples->nbins()) return;
+  if (tuples->size(idx)==0) {
+    return;
+  }
+
+  quality[idx] = pixelTuplesHeterogeneousProduct::bad;
+
+  // only quadruplets
+  if (tuples->size(idx)<4) { 
+    return;
+  }
+
+  bool isNaN = false;
+  for (int i=0; i<5; ++i) {
+    isNaN |=  fit_results[idx].par(i)!=fit_results[idx].par(i);
+  }
+  isNaN |=  !(fit_results[idx].chi2_line+fit_results[idx].chi2_circle < 100.f);  // catch NaN as well
+
+  // impose "region cuts" (NaN safe)
+  // phi,Tip,pt,cotan(theta)),Zip
+  bool ok = std::abs(fit_results[idx].par(1)) < 0.1f 
+         && fit_results[idx].par(2) > 0.3f
+         && std::abs(fit_results[idx].par(4)) < 12.f;
+  ok &= (!isNaN);
+  quality[idx] = ok ? pixelTuplesHeterogeneousProduct::loose : pixelTuplesHeterogeneousProduct::bad; 
 }
 
 __global__
-void kernelPrintFoundNtuplets(GPU::SimpleVector<Quadruplet> *foundNtuplets, int maxPrint)
-{
+void kernel_print_found_ntuplets(TuplesOnGPU::Container * foundNtuplets, uint32_t maxPrint) {
   for (int i = 0; i < std::min(maxPrint, foundNtuplets->size()); ++i) {
     printf("\nquadruplet %d: %d %d %d %d\n", i,
-           (*foundNtuplets)[i].hitId[0],
-           (*foundNtuplets)[i].hitId[1],
-           (*foundNtuplets)[i].hitId[2],
-           (*foundNtuplets)[i].hitId[3]
+           (*(*foundNtuplets).begin(i)),
+           (*(*foundNtuplets).begin(i)+1),
+           (*(*foundNtuplets).begin(i)+2),
+           (*(*foundNtuplets).begin(i)+3)
           );
-
   }
 }
 
-void wrapperPrintFoundNtuplets(
-    int gridSize, int blockSize, cudaStream_t cudaStream,
-    GPU::SimpleVector<Quadruplet> *foundNtuplets,
-    int maxPrint)
+void CAHitQuadrupletGeneratorKernels::launchKernels( // here goes algoparms....
+                                                HitsOnCPU const & hh,
+                                                TuplesOnGPU & tuples_d,
+                                                cudaStream_t cudaStream)
 {
-  kernelPrintFoundNtuplets<<<gridSize, blockSize, 0, cudaStream>>>(
-    foundNtuplets,
-    maxPrint);
+  auto & gpu_ = tuples_d;
+  auto maxNumberOfDoublets_ = CAConstants::maxNumberOfDoublets();
+
+
+  auto nhits = hh.nHits;
+  assert(nhits <= PixelGPUConstants::maxNumberOfHits);
+  auto blockSize = 64;
+  auto stride = 4;
+  auto numberOfBlocks = (nhits + blockSize - 1)/blockSize;
+  numberOfBlocks *=stride;
+  fishbone<<<numberOfBlocks, blockSize, 0, cudaStream>>>(
```

Comment:
for a fast test one can comment this kernel to see if go faster, duplicates will increase, not beyond previous version.